crates/markdown_preview/src/markdown_preview_view.rs

```diff
@@ -0,0 +1,166 @@
+    InteractiveElement, IntoElement, ListState, ParentElement, Render, Styled, View, ViewContext,
+};
+use language::LanguageRegistry;
+use std::sync::Arc;
+use ui::prelude::*;
+use workspace::item::Item;
+use workspace::Workspace;
+
+use crate::OpenPreview;
+
+pub struct MarkdownPreviewView {
+    focus_handle: FocusHandle,
+    list_state: ListState,
+}
+
+impl MarkdownPreviewView {
+    pub fn register(workspace: &mut Workspace, _cx: &mut ViewContext<Workspace>) {
+        let languages = workspace.app_state().languages.clone();
+
+        workspace.register_action(move |workspace, _: &OpenPreview, cx| {
+            if workspace.has_active_modal(cx) {
+                cx.propagate();
+                return;
+            }
+            let languages = languages.clone();
+            let editor = workspace.active_item_as::<Editor>(cx).unwrap();
+            MarkdownPreviewView::deploy_preview(workspace, languages, editor, cx);
+            cx.notify();
+        });
+    }
+
+    pub fn new(
+        active_editor: View<Editor>,
+        languages: Arc<LanguageRegistry>,
+        cx: &mut ViewContext<Self>,
+    ) -> Self {
+        let focus_handle = cx.focus_handle();
+
+        cx.subscribe(&active_editor, |_this, _editor, event: &EditorEvent, cx| {
+            if *event == EditorEvent::Edited {
+                cx.notify();
+            }
+        })
+        .detach();
+
+        let list_state = ListState::new(1, gpui::ListAlignment::Top, px(1000.), move |_ix, cx| {
+            let editor = active_editor.read(cx);
+            let contents = editor.buffer().read(cx).snapshot(cx).text();
+
+            let mentions = vec![];
+            let text = rich_text::render_markdown(contents, &mentions, &languages, None);
+
+            v_flex()
+                .w_full()
+                .id("markdown_preview_container")
+                .group("")
+                .child(text.element("body".into(), cx))
+                .into_any()
+        });
+
+        Self {
+            list_state,
+            focus_handle,
+        }
+    }
+
+    // Re-active the most recently active preview, or create a new one if there isn't one.
+    fn deploy_preview(
+        workspace: &mut Workspace,
+        languages: Arc<LanguageRegistry>,
+        active_editor: View<Editor>,
+        cx: &mut ViewContext<Workspace>,
+    ) {
+        let existing = workspace
+            .active_pane()
+            .read(cx)
+            .items()
+            .find_map(|item| item.downcast::<MarkdownPreviewView>());
+
+        Self::existing_or_new_preview(workspace, languages, active_editor, existing, cx)
+    }
+
+    fn existing_or_new_preview(
+        workspace: &mut Workspace,
+        languages: Arc<LanguageRegistry>,
+        active_editor: View<Editor>,
+        existing: Option<View<MarkdownPreviewView>>,
+        cx: &mut ViewContext<Workspace>,
+    ) {
+        if let Some(existing) = existing {
+            workspace.activate_item(&existing, cx);
```

Comment:
We should probably check that the new preview corresponds to the editor that this action was run on, and if not, replace it's state.

Reply:
Can you point me to somewhere in the codebase that does something similar? I'm struggling to figure this one out.